pkg/agent/cniserver/ipam/antrea_ipam_controller.go

```diff
@@ -81,21 +117,100 @@ func (c *AntreaIPAMController) Run(stopCh <-chan struct{}) {
 	}()
 
 	klog.InfoS("Starting", "controller", controllerName)
-	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced) {
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.namespaceInformer.Informer().HasSynced, c.ipPoolInformer.Informer().HasSynced, c.podInformer.HasSynced) {
 		return
 	}
 
 	<-stopCh
 }
 
-func (c *AntreaIPAMController) getIPPoolsByNamespace(namespace string) []string {
+func (c *AntreaIPAMController) getIPPoolsByPod(namespace, name string) ([]string, []net.IP, *crdv1a2.IPAddressOwner, error) {
+	// Find IPPool by Pod
+	var IPs []net.IP
+	var reservedOwner *crdv1a2.IPAddressOwner
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	if err != nil {
+		// When get Pod failed, try to get information from allocated IPs of all IPPools.
+		klog.Infof("Get pod %s/%s failed, err=%+v", namespace, name, err)
+		ipPools, _ := c.ipPoolInformer.Informer().GetIndexer().ByIndex(podIndex, k8s.NamespacedName(namespace, name))
+		for _, item := range ipPools {
+			ipPool := item.(*crdv1a2.IPPool)
+			if ipPool.Spec.IPVersion != 4 {
+				continue
+			}
+			for _, IPAddress := range ipPool.Status.IPAddresses {
+				if IPAddress.Owner.Pod != nil && IPAddress.Owner.Pod.Namespace == namespace && IPAddress.Owner.Pod.Name == name {
+					if IPAddress.Owner.StatefulSet != nil {
+						reservedOwner = &crdv1a2.IPAddressOwner{StatefulSet: IPAddress.Owner.StatefulSet}
+					}
+					return []string{ipPool.Name}, []net.IP{net.ParseIP(IPAddress.IPAddress)}, reservedOwner, nil
+				}
+			}
+		}
+		return nil, nil, nil, err
+	}
+	// Collect specified IPs if exist
+	IPStrings, _ := pod.Annotations[AntreaIPAMPodIPAnnotationKey]
+	IPStrings = strings.ReplaceAll(IPStrings, " ", "")
+	var ipErr error
+	if IPStrings != "" {
+		splittedIPStrings := strings.Split(IPStrings, AntreaIPAMAnnotationDelimiter)
+		for _, IPString := range splittedIPStrings {
+			IP := net.ParseIP(IPString)
+			if IPString != "" && IP == nil {
+				ipErr = fmt.Errorf("invalid IP annotation %s", IPStrings)
+				IPs = nil
+				break
+			}
+			IPs = append(IPs, IP)
+		}
+	}
+
+ownerReferenceLoop:
+	for _, ownerReference := range pod.OwnerReferences {
+		if ownerReference.Controller != nil && *ownerReference.Controller == true {
+			switch ownerReference.Kind {
+			case "StatefulSet":
+				statefulSetName, index, err := k8s.ParseStatefulSetName(name)
+				if err != nil {
+					klog.Warningf("Invalid StatefulSet name: %s", name)
+					break ownerReferenceLoop
```

Comment:
I think better to add some comments here.

Reply:
Sure.

Reply:
I meant why we break to ownerReferenceLoop here.

Reply:
The pod Name is not a StatefulSet Pod name, and we don't need to process other `OwnerReferences`.
Comments added.
`// This should not occur unless user creates an invalid Pod manually`